### PR TITLE
Changes to Sapling bundle mapping

### PR DESCRIPTION
**Pull request by str4d**
Also includes a few new `Clone` and `Debug` impls we need.

Part of #741.

### CODE REVIEW

**Comment by str4d** [comment]
The tuple impls let me write code like:
```rust
impl Bundle<InProgress<Proven, PartiallyAuthorized>> {
    /// Finalizes this bundle, enabling it to be included in a transaction.
    ///
    /// Returns an error if any signatures are missing.
    pub fn finalize(self) -> Result<Bundle<Authorized>, Error> {
        self.try_map_authorization((
            |p| Ok(p),
            |p| Ok(p),
            |maybe: MaybeSigned| maybe.finalize(),
            |partial: InProgress<Proven, PartiallyAuthorized>| {
                Ok(Authorized {
                    binding_sig: partial.sigs.binding_signature,
                })
            },
        ))
    }
}
```
which has the same usability as the `orchard` crate's implementations, while still allowing the maps to be defined as structs if desired.

**Comment by str4d** [comment]
Meanwhile, having the `MapAuth` trait methods take `&mut self` means I can write code like:
```rust

impl<P: InProgressProofs> Bundle<InProgress<P, Unsigned>> {
    /// Loads the sighash into this bundle, preparing it for signing.
    ///
    /// This API ensures that all signatures are created over the same sighash.
    pub fn prepare<R: RngCore + CryptoRng>(
        self,
        mut rng: R,
        sighash: [u8; 32],
    ) -> Bundle<InProgress<P, PartiallyAuthorized>> {
        self.map_authorization((
            |proof| proof,
            |proof| proof,
            |parts| MaybeSigned::SigningMetadata(parts),
            |auth: InProgress<P, Unsigned>| InProgress {
                sigs: PartiallyAuthorized {
                    binding_signature: auth.sigs.bsk.sign(
                        &sighash,
                        &mut rng,
                        VALUE_COMMITMENT_RANDOMNESS_GENERATOR,
                    ),
                    sighash,
                },
                _proof_state: PhantomData::default(),
            },
        ))
    }
}
```
which requires `FnMut` closures in order to access `rng` (or equivalently `&mut self` to update mutable fields of struct-based maps).

**Review by daira — APPROVED** [review]
utACK

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 1]
```diff
@@ -56,11 +56,14 @@ impl Authorization for Authorized {
...
+/// For use with [`Bundle::map_authorization`].
 pub trait MapAuth<A: Authorization, B: Authorization> {
-    fn map_spend_proof(&self, p: A::SpendProof) -> B::SpendProof;
-    fn map_output_proof(&self, p: A::OutputProof) -> B::OutputProof;
-    fn map_auth_sig(&self, s: A::AuthSig) -> B::AuthSig;
-    fn map_authorization(&self, a: A) -> B;
+    fn map_spend_proof(&mut self, p: A::SpendProof) -> B::SpendProof;
+    fn map_output_proof(&mut self, p: A::OutputProof) -> B::OutputProof;
+    fn map_auth_sig(&mut self, s: A::AuthSig) -> B::AuthSig;
+    fn map_authorization(&mut self, a: A) -> B;
```
Comment (nuttycom) [inline zcash_primitives/src/transaction/components/sapling.rs:66]: `map` with a `&mut self` reference seems increasingly misnamed, and the commit text doesn't explain the rationale for this change, so I'm assuming I'll be able to figure it out from later commits.

**Review by nuttycom — APPROVED** [review]
utACK modulo documentation; it's not clear at all why it's necessary to have the transformation algebra itself be mutable.

EDIT: I see now the PR comments.

File: zcash_primitives/src/transaction/components/sapling.rs (hunk 1 above)
Comment (str4d) [reply zcash_primitives/src/transaction/components/sapling.rs:66]: I gave an example of why I needed this in a comment on the PR.

**Pull request merged**